tools/toolchain/clang_lto_pgo.sh

```diff
@@ -0,0 +1,76 @@
+cd stage-0
+
+cmake -B build -S llvm -DLLVM_ENABLE_PROJECTS='clang;lld;bolt' -DLLVM_ENABLE_RUNTIMES='compiler-rt' -G Ninja -DLLVM_TARGETS_TO_BUILD=X86 -DCMAKE_BUILD_TYPE=Release -DCOMPILER_RT_BUILD_SANITIZERS=Off -DCOMPILER_RT_DEFAULT_TARGET_ONLY=On
+cmake --build build --
+
+USE_NEW_COMPILER=(-DCMAKE_C_COMPILER="$DIR/stage-0/build/bin/clang" -DCMAKE_CXX_COMPILER="$DIR/stage-0/build/bin/clang++" -DLLVM_USE_LINKER="$DIR/stage-0/build/bin/ld.lld")
+COMMON_OPTS=(-DLLVM_TARGETS_TO_BUILD=X86 -DLLVM_ENABLE_LTO=Thin -DLLVM_BUILD_RUNTIME=No -GNinja -DCLANG_DEFAULT_PIE_ON_LINUX=OFF)
+
+# Build a PGO-optimized compiler using the boostrapped compiler.
+# Choose a version compatible with your system.
+# If you change it, remember to change clang-15 in the later part of the script to the appropriate name as well.
+git fetch --depth=1 origin tag llvmorg-15.0.6
+git worktree add ../stage-1 llvmorg-15.0.6
+cd "../stage-1"
+cmake -B build -S llvm -DLLVM_ENABLE_PROJECTS='clang' -DCMAKE_BUILD_TYPE=Release "${USE_NEW_COMPILER[@]}" "${COMMON_OPTS[@]}" -DLLVM_BUILD_INSTRUMENTED=IR
+cmake --build build -- bin/clang
+
+cd ../scylla
+rm -rf build build.ninja
+./configure.py --mode=release --compiler=$(realpath ../stage-1/build/bin)/clang++ --disable-dpdk
+LLVM_PROFILE_FILE=$(realpath ../stage-1)/build/profiles/ir-%p-%m.profraw ninja build/release/scylla
+
+cd ../stage-1
+../stage-0/build/bin/llvm-profdata merge build/profiles/ir-*.profraw -output=ir.prof
+rm -r build
+cmake -B build -S llvm -DLLVM_ENABLE_PROJECTS='clang' -DCMAKE_BUILD_TYPE=Release "${USE_NEW_COMPILER[@]}" "${COMMON_OPTS[@]}" -DLLVM_BUILD_INSTRUMENTED=CSIR -DLLVM_PROFDATA_FILE=$(realpath ir.prof)
+cmake --build build -- bin/clang
+
+cd ../scylla
+rm -rf build build.ninja
+./configure.py --mode=release --compiler=$(realpath ../stage-1/build/bin)/clang++ --disable-dpdk
+LLVM_PROFILE_FILE=$(realpath ../stage-1)/build/profiles/csir-%p-%m.profraw ninja build/release/scylla
+
+cd ../stage-1
+../stage-0/build/bin/llvm-profdata merge build/profiles/csir-*.profraw -output=csir.prof
+../stage-0/build/bin/llvm-profdata merge ir.prof csir.prof -output=combined.prof
+rm -r build
+# -DLLVM_LIBDIR_SUFFIX=64 for Fedora compatibility
+cmake -B build -S llvm -DLLVM_ENABLE_PROJECTS='clang' -DCMAKE_BUILD_TYPE=Release "${USE_NEW_COMPILER[@]}" "${COMMON_OPTS[@]}" -DLLVM_PROFDATA_FILE=$(realpath combined.prof) -DCMAKE_EXE_LINKER_FLAGS="-Wl,--emit-relocs" -DCMAKE_INSTALL_PREFIX=/usr -DLLVM_LIBDIR_SUFFIX=64
+cmake --build build -- bin/clang
+
+mv build/bin/clang-15 build/bin/clang-15.prebolt
+mkdir build/profiles
+../stage-0/build/bin/llvm-bolt build/bin/clang-15.prebolt -o build/bin/clang-15 -instrument -instrumentation-file=$(realpath build/profiles)/prof -instrumentation-file-append-pid -conservative-instrumentation
+
+cd ../scylla
+rm -rf build build.ninja
+./configure.py --mode=release --compiler=$(realpath ../stage-1/build/bin)/clang++ --disable-dpdk
```

Comment:
We'll want to train on dev too, since that's what most people use daily.

Reply:
I'd expect it to make 0 difference. Aren't the only real difference between `dev` and `release` that `dev` doesn't produce debug info, and doesn't increase the inline threshold?

Reply:
-O1 vs -O3

Reply:
But, probably it won't make a difference, since -O3 is a superset of -O1.

Reply:
`dev` is O2, not O1. O1 would be crippling for testing because, until recently, it didn't do inlining.

And the appearance of inlining was the biggest difference between O1, O2 and O3. Nowadays, since that difference is gone, the difference is at follows:

O1 has almost all optimizations
O2 adds jump threading, tail call elimination, merged load/store motion (hoisting common stores outside both branches of `if`), dead store elimination and SLP vectorizer.
O3 adds call site splitting, and pointer->value argument promotion.

So the differences between optimization levels are very small nowadays — only a few passes out of a hundred. And as you noted, the higher levels are supersets of the higher ones — so I don't see a point training on the lower levels.

Reply:
Ah, let's switch dev to -O1 then. Agree no need to train on dev.

Reply:
(switching dev to -O1 outside the scope here, of course)